### ISSUE TITLE
youtube.videos.insert can't attach media body

### ISSUE THREAD

**Issue by tomfox747** [google/google-api-javascript-client#625]
I've been trying to upload a youtube video using the javascript client library. I have all of the authentication code working correctly and i've been able to call youtube.channels.list successfully.

However when I attempt to call youtube.videos.insert I keep hitting the following error 
(400, 'mediaBodyRequired', 'the request does not include the video content'). I can't seem to find any information on how to attach the video data to the request. I've tried various data formats including binary, blobs and files but keep getting the same error.

Can someone provide an example of how this needs to be done?

**Comment by yingted** [comment]
Reviving a dead issue.

I had a similiar issue. It looks like the JS client doesn't support multi-part uploads.

Neither of these pages show the "EXECUTE" button, but `/list` does for both.
https://developers.google.com/youtube/v3/docs/videos/insert
https://developers.google.com/youtube/v3/docs/captions/update

**Comment by yingted** [comment]
Okay, it looks like the discovery generated methods don't accept a media body parameter.

I set a breakpoint in the Python example code, printed the request, and copy pasted it to the javascript code.

```js
gapi.client.request({
  path: 'https://www.googleapis.com/upload/youtube/v3/captions',
  method: 'POST',
  params: { part: 'snippet', alt: 'json', uploadType: 'multipart', },
  headers: { accept: 'application/json', 'content-type': 'multipart/related; boundary="random123"', },
  body: '--random123\nContent-Type: application/json\nMIME-Version: 1.0\n\n' + JSON.stringify({snippet: snippet}) + '\n--random123\nContent-Type: application/octet-stream\nMIME-Version: 1.0\nContent-Transfer-Encoding: binary\n\n' + uploadContents + '\n--boundary--\n',
})
```

It seems to work so far, but I've exceeded my quota so I will have to come back later.

**Comment by tomlarkworthy** [comment]
the ending "--boundary--" should be "--random123--"
I have it uploading to me account!!! However, the video is corrupt and does not play. (body encoding issue??)

**Comment by tomlarkworthy** [comment]
YaaaaaaaaaS!!!! https://observablehq.com/@tomlarkworthy/youtube-upload
It was problems doing the binary encoding that messed me up. I changed to base64 and it worked!!!
https://www.youtube.com/watch?v=FThG8pVGIS0&feature=youtu.be

```
gapi.client.request({
  path: 'https://www.googleapis.com/upload/youtube/v3/videos',
  method: 'POST',
  params: { part: 'id,snippet,status', alt: 'json', uploadType: 'multipart'},
  headers: { accept: 'application/json', 'content-type': 'multipart/related; boundary="random123"', },
  body: '--random123\nContent-Type: application/json\nMIME-Version: 1.0\n\n' + JSON.stringify({
    snippet: {
          title:"test video 3",
          description:"this is a test video"
        },
        status:{
          privacyStatus:'unlisted'
        }
      }
  ) + '\n--random123\nContent-Type: video/mp4\nMIME-Version: 1.0\nContent-Transfer-Encoding: base64\n\n' + btoa(Uint8ToString(contents)) + '\n--random123--\n',
})
```

**Comment by tp27933** [comment]
@tomlarkworthy where did you put your video source in the code? is the variable named contents?  I am using input tag so I get the file by accessing e.target.file[0]. In my case , how do i convert the  file ?
I am newbie in Javascript ,many thanks in advance!!!

**Comment by tomlarkworthy** [comment]
yeah, the variable contents is refering to a cell further up, which had its source code closed in the hamburger menu which is hard to follow if you are not used to Observable. Anyway, the code for contents and the Uint8ToString encoder is 

<img width="1223" alt="image" src="https://user-images.githubusercontent.com/1848162/185162687-3d13bb1b-83e8-49b4-a251-cee4df46771f.png">

I pinned them open so you can copy and paste from the notebook now.

It seems like I could probably do a simpler set of transforms from binary data to the base encoded version, but as each failed attempt consumes quota its very hard to iterate so I will leave it as it is.

**Comment by LowCodeInvestigater** [comment]
Hey @tomlarkworthy I want to achive the some approach with LowCode but I have only the capability to send a HTTP request. Could you give me an hint how to convert your body in a proper form? 

![image](https://user-images.githubusercontent.com/122006272/210726082-727264ad-950b-4c0c-ba86-5f3fdc18a27d.png)

**Comment by tp27933** [comment]
> yeah, the variable contents is refering to a cell further up, which had its source code closed in the hamburger menu which is hard to follow if you are not used to Observable. Anyway, the code for contents and the Uint8ToString encoder is
> 
> <img alt="image" width="1223" src="https://user-images.githubusercontent.com/1848162/185162687-3d13bb1b-83e8-49b4-a251-cee4df46771f.png">
> 
> I pinned them open so you can copy and paste from the notebook now.
> 
> It seems like I could probably do a simpler set of transforms from binary data to the base encoded version, but as each failed attempt consumes quota its very hard to iterate so I will leave it as it is.

Thanks , you really saved my day

**Comment by tomlarkworthy** [comment]
@LowCodeInvestigater  so the multipart body has two main payloads, the YouTube JSON message, which is just some stringification into a staticically fixed JSON schema, which most low code tools allow string interpolation so maybe its a pain but probably possible. The other part is rendering the file content  nto a base64 encoded payload. This might be hard to do as you need to read the file content and do some processing, and this really depends on the tool and how it exposes file uploads (if at all). You have not given enough context to determine this but I would expect its probably not possible in a low code env.

**Comment by LowCodeInvestigater** [comment]
@tomlarkworthy actually I just need to know whether its possible send a Body like in the image. I am working with MS Power Automate and for now, i was able to handle each file content. 
![image](https://user-images.githubusercontent.com/122006272/210730339-0a8c2071-66d4-4ac0-92f1-673a52cd81fc.png)


![image](https://user-images.githubusercontent.com/122006272/210729816-94841917-d8ec-4c72-aa53-421bb002651c.png)

**Comment by tomlarkworthy** [comment]
I think the easiest way for you to experience how this works is to clone my notebook, get it working on your own Google account account, and then look at the network tab to *see* what a working solution looks like on the wire, and then try to convert the working solution to your low code tool. Ultimately on the wire this is a single request with a text body, so any generic HTTP handler can issue it, and there is just a bit of complexity on how to generate the body dynamically based on real file contents. But you could copy and pase a working body from the network tab and slowly convert a static solution to a dynamic one once you have the static version cloned in your own tool.

**Comment by LowCodeInvestigater** [comment]
Thanks for your quick reply, it helps me a lot! I will give it a try and in best case I will post a solution :)

**Comment by LowCodeInvestigater** [comment]
It looks  very closer but now I receive an error and I need to admit that I am not so familar with javascript. If I post my request I receive a Bad Request due to this erros Message "Invalid multipart request with 0 mime parts.". Do you know how to handle this error? 

![image](https://user-images.githubusercontent.com/122006272/210748858-a6e5c53c-ba99-4c5f-8739-f6200f4796d5.png)

**Comment by tomlarkworthy** [comment]
oh yeah, you have quoted your payload with single quotes, thats for JS and is not part of the body content. Also the `\n` are represent newlines, not concrete characters. I think it is the lack of newlines that means the multipart payload is not recognisably split into segments

**Comment by LowCodeInvestigator** [comment]
I am getting closer but now I am out of quota LOL.. I will get a response upon my request and I see the metadata has been patached. But I it seems something is wrong with my base 64 content. Lets see how many days it takes..

**Comment by LowCodeInvestigator** [comment]
Solved but with a different approach, for now I have to send two request. But it works very well!  Do you know a good site for more details about the multipart? I would like to dig in

**Comment by tomlarkworthy** [comment]
Multipart is a generic HTTP mechanism for stuffing multiple distinct http payloads into a single request https://stackoverflow.com/questions/16958448/what-is-http-multipart-request so you can see here we are sending a JSON payload and a file payload.
